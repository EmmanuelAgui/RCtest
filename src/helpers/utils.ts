export function isEmpty(str: string | undefined):boolean{
    return !str;
}

export function isNumber(n:any):boolean {
    return !isNaN(parseFloat(n)) && isFinite(n);
}