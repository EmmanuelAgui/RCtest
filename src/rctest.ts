import { Sequence } from "./index";
// @ts-ignore
export let sequence = new Sequence();
console.log(sequence.next());
