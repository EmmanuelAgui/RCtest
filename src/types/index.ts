export enum EXT_TYPE{
    DigitalUser = 'DigitalUser',
    VirtualUser = 'VirtualUser',
    FaxUser = 'VirtualUser',
    AO = 'AO',
    Dept='Dept'
  }

export interface Extension {
    firstName: string
    lastName?: string
    ext?: string
    extType: EXT_TYPE
    fullName?: string // "firstName" + "lastName" + "ext" 
    extTypeNum?:number // map from extType
    [prop:string]:any
}
  // extensions function type
export type ExtensionFun = (extensions: Extension[], sortKey?: keyof Extension, sortType?: 'ASC' | 'DESC') => Extension[];

export interface SaleItem{
    month: number
    date: number
    transationId: string
    salePrice: number
}
  
export interface QuarterItem{
    quarter: number
    transactionNums: number
}
  
export interface QuarterSumItem extends QuarterItem{
    totalPrices: number
}
  
export interface QuarterAverageItem extends QuarterItem{
    averagePrices: number
}