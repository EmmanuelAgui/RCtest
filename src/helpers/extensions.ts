import { ExtensionFun, Extension } from "../types";
import { isEmpty, isNumber } from "./utils";

export function setFullNameOfExtensions(extensions: Extension[]): Extension[]{
    let extensionsWithFullName = extensions.map(item => {
      item.fullName = getFullName(item);
      return item;
    })
    return extensionsWithFullName;
    // get fullName of an extension
    function getFullName(item:Extension):string{
      let fullName =  item.firstName
      !isEmpty(item.lastName)&&(fullName+=item.lastName)
      !isEmpty(item.ext)&&(fullName+=item.ext)
      return fullName
    }
  }
  
export function setExtTypeNumOfExtensions (extensions: Extension[]): Extension[]{
    let extensionsWithExtTypeNum = extensions.map(item => {
      item.extTypeNum = getExtTypeNum(item);
      return item;
    })
    return extensionsWithExtTypeNum;
  
    function getExtTypeNum(item: Extension):number{
      let mapArray: string[] = ["DigitalUser", "VirtualUser", "FaxUser", "AO","Dept"];
      return mapArray.indexOf(item.extType);
    }
  }
  
export function sortExtensionsByKey(extensions: Extension[], sortKey: keyof Extension, sortType:string = 'ASC'): Extension[]{ 
    extensions.sort(function (a, b) {
        if (isNumber(a[sortKey]) && isNumber(b[sortKey])) {
        return sortType === 'DESC' ? b[sortKey] - a[sortKey] : a[sortKey] - b[sortKey];
        } else {
        return sortType === 'DESC' ? b[sortKey].localeCompare(a[sortKey]) : a[sortKey].localeCompare(b[sortKey]);
        } 
    });
    return extensions;
}
  
export function getSortedExtensions(extensions: Extension[]): Extension[]{ 
    let result = extensions.map(item => {
        const {firstName,lastName,ext,extType }: Extension = item;
        let extensionItem:Extension = { firstName, lastName, ext, extType };
        return extensionItem;
    })
    return result
}

