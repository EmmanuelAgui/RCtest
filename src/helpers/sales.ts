import { SaleItem, QuarterSumItem, QuarterAverageItem } from "../types";

export function sortSaleItems(saleItems: SaleItem[]): SaleItem[] {
    saleItems.sort((a, b) => a.month - b.month)
    return saleItems
}

export function getQuarterFromMonth(month: number): number {
    if (month > 12 || month < 1) {
        throw Error('Please input a correct month');
    }
    if (month > 0 && month <= 3) {
        return 1;
    } else if (month > 3 && month <= 6) {
        return 2;
    } else if (month > 6 && month <= 9) {
        return 3;
    } else {
        return 4;
    }
}

export function initQuarterSumItem(quarter: number, quarterSumItems: QuarterSumItem[]): QuarterSumItem {
    let quarterSumItem = {
        quarter: quarter,
        totalPrices: 0,
        transactionNums: 0
    }
    quarterSumItems.push(quarterSumItem)
    return quarterSumItem;
}

export function getQuarterSumItem(quarter: number, quarterSumItems: QuarterSumItem[]): QuarterSumItem {
    // @ts-ignore
    let quarterSumItem = quarterSumItems.find(item => item.quarter === quarter);
    !quarterSumItem && (quarterSumItem = initQuarterSumItem(quarter, quarterSumItems));
    return quarterSumItem as QuarterSumItem;
}

export function sumSalesByQuarter(saleItems: SaleItem[]): QuarterSumItem[] {
    let result: QuarterSumItem[] = [];
    saleItems.reduce((quarterSumItems, currentValue) => {
        // get quarter n
        let quarter = getQuarterFromMonth(currentValue.month);
        // find or create quarterSumItem n
        let quarterSumItem = getQuarterSumItem(quarter, quarterSumItems);
        // accumulate the totalPrices and transactionNums of quarterSumItem n
        quarterSumItem.totalPrices += currentValue.salePrice;
        quarterSumItem.transactionNums++;
        // return quarter n
        return quarterSumItems
    }, result)
    return result;
}

export function getQuarterAverageFrom(sum: QuarterSumItem[]): QuarterAverageItem[] {
    let result: QuarterAverageItem[];
    result = sum.map(item => {
        let quarterAverageItem: QuarterAverageItem;
        let averagePrices = Number(item.totalPrices / item.transactionNums).toFixed(2);
        quarterAverageItem = {
            quarter: item.quarter,
            averagePrices: Number(averagePrices),
            transactionNums: item.transactionNums
        };
        return quarterAverageItem;
    })
    return result;
}
