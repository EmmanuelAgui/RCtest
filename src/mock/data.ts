import { EXT_TYPE } from "../types";

// test data
export let myExtensions = [
    { firstName: 'adfsdf', lastName: 'xxddx', ext: 'adfx', extType: EXT_TYPE.AO },
    { firstName: 'rtfgh', lastName: 'sfdf', ext: '', extType: EXT_TYPE.FaxUser },
    { firstName: 'dfdf', lastName: 'dfs', ext: 'adfx', extType: EXT_TYPE.Dept },
    { firstName: 'dfd', lastName: '', ext: 'adfx', extType: EXT_TYPE.Dept },
    { firstName: 'sfwf', lastName: 'ffdfg', ext: '', extType: EXT_TYPE.VirtualUser },
    { firstName: 'reer', lastName: 'dfg', ext: '', extType: EXT_TYPE.DigitalUser },
    { firstName: 'cvfg', lastName: 'hts', ext: 'adfx', extType: EXT_TYPE.AO },
    { firstName: 'bwerew', lastName: '', ext: 'adfx', extType: EXT_TYPE.FaxUser },
    { firstName: 'weewr', lastName: 'derter', ext: 'adfx', extType: EXT_TYPE.Dept }
];
export let mySaleItems = [
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 4,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 5,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 9,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 7,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 1,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    },
    {
        month: 5,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    }, {
        month: 11,
        date: 3,
        transationId: "sdfsdf",
        salePrice: 12
    }
];
  