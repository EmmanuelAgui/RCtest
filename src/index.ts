import { ExtensionFun, Extension, EXT_TYPE, SaleItem, QuarterSumItem, QuarterAverageItem } from "./types";
import { setFullNameOfExtensions, sortExtensionsByKey, getSortedExtensions, setExtTypeNumOfExtensions } from "./helpers/extensions";
import { sortSaleItems, sumSalesByQuarter, getQuarterAverageFrom } from "./helpers/sales";
import { myExtensions, mySaleItems } from "./mock/data";
  
/**
    extensions is an Array and each item has such format:
    {firstName: 'xxx', lastName: 'xxx', ext: 'xxx', extType: 'xxx'}
    lastName, ext can be empty, extType can only has "DigitalUser", "VirtualUser","FaxUser","Dept","AO".
**/
  
/**
    Question 1: sort extensions by "firstName" + "lastName" + "ext" ASC
**/
export let sortExtensionsByName: ExtensionFun = (extensions: Extension[]): Extension[] => {
    // extend fullName ("firstName" + "lastName" + "ext") of extension item
    let extensionsWithFullName = setFullNameOfExtensions(extensions);
    // sort extensions by fullName
    let sortedExtensionsWithFullName = sortExtensionsByKey(extensionsWithFullName, 'fullName');
    // get result extensions
    extensions = getSortedExtensions(sortedExtensionsWithFullName);
    return extensions;
}

/**
    Question 2: sort extensions by extType follow these orders ASC
    DigitalUser < VirtualUser < FaxUser < AO < Dept.
**/
export let sortExtensionsByExtType: ExtensionFun = (extensions: Extension[]): Extension[] => {
    // extend extTypeNum of extension item
    let extensionsWithExtTypeNum = setExtTypeNumOfExtensions(extensions);
    // sort extensions by extTypeNum
    let sortedExtensionsWithExtTypeNum = sortExtensionsByKey(extensionsWithExtTypeNum, 'extTypeNum');
    // get result extensions
    extensions = getSortedExtensions(sortedExtensionsWithExtTypeNum);
    return extensions;
}
// // test
// let test = sortExtensionsByName(myExtensions);
// console.log(test);
// test = sortExtensionsByExtType(myExtensions);
// console.log(test);
/**
    saleItems is an Array has each item has such format:
    {
        month: n, //[1-12],
        date: n, //[1-31],
        transationId: "xxx",
        salePrice: number
    }
**/

/**
    Question 3: write a function to calculate and return a list of total sales (sum) for each quarter, expected result like:
    [
        {quarter: 1, totalPrices: xxx, transactionNums: n},
        {....}
    ]
**/
export function sumByQuarter(saleItems: SaleItem[]): QuarterSumItem[]{
    let result: QuarterSumItem[]
    // sort saleItems by month
    saleItems = sortSaleItems(saleItems);
    // calcuate totalPrices and transactionNums of each quarter
    result = sumSalesByQuarter(saleItems);
    return result;
}
/**
    Question 4: write a function to calculate and return a list of average sales for each quarter, expected result like:
    [
        {quarter: 1, averagePrices: xxx, transactionNums: n},
        {....}
    ]
**/

export function averageByQuarter(saleItems: SaleItem[]): QuarterAverageItem[]{
    let result: QuarterAverageItem[];
    // calcuate totalPrices and transactionNums of each quarter
    let sumResult: QuarterSumItem[] = sumByQuarter(saleItems);
    // calcuate averagePrices of each quarter
    result = getQuarterAverageFrom(sumResult);
    return result;
}
//test
let quarterSumItems = sumByQuarter(mySaleItems);
console.log(quarterSumItems);
let quarterAverageItems = averageByQuarter(mySaleItems);
console.log(quarterAverageItems);
  
/**
    Question 5: please create a tool to generate Sequence
    Expected to be used like:
    var sequence1 = new Sequence();
    sequence1.next() --> return 1;
    sequence1.next() --> return 2;
    
    in another module:
    var sequence2 = new Sequence();
    sequence2.next() --> 3;
    sequence2.next() --> 4;
**/
// js Singleton
export const Sequence = (function () {
    // store instance
    let _instance:any=null;

    const Sequence = function () {
    if (_instance) return _instance;// return the instance exist
    // @ts-ignore
    _instance = this;
    // @ts-ignore
    this.init();
    return _instance;
    }

    Sequence.prototype.init = function () {
    _instance.source = 0;
    }
    Sequence.prototype.next = function () {
    return ++_instance.source;
    }

    return Sequence
})()
// @ts-ignore
const s1 = new Sequence();
// @ts-ignore
const s2 = new Sequence();
console.log(s1.next());
console.log(s2.next());
console.log(s1.next());

/**
     Question 6:
    AllKeys: 0-9;
    usedKeys: an array to store all used keys like [2,3,4];
    We want to get an array which contains all the unused keys,in this example it would be: [0,1,5,6,7,8,9]
**/

export function getUnUsedKeys<T>(allKeys:T[], usedKeys:T[]):T[]{
    // @ts-ignore
    let unUsedKeys: T[] = allKeys.filter(key => !usedKeys.includes(key));
    return unUsedKeys;
}
// test
// let allKeys = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
// let usedKeys = [2, 3, 4];
// let unUsedKeys = getUnUsedKeys(allKeys, usedKeys);
// console.log(unUsedKeys);
  