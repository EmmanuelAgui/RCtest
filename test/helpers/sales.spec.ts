import { sortSaleItems, getQuarterFromMonth, getQuarterSumItem, sumSalesByQuarter, getQuarterAverageFrom, initQuarterSumItem } from '../../src/helpers/sales';
import { mySaleItems } from '../../src/mock/data';
import { QuarterSumItem, QuarterAverageItem } from '../../src/types';

describe('helpers:sales', () => {
    let quarterSumItems: QuarterSumItem[];
    describe('get sales sorted by month', () => {
        test('should get sales sorted by a specified key ascend', () => {
            let sortedSales = sortSaleItems(mySaleItems);
            expect(sortedSales).toMatchSnapshot();
        })
    })

    test('should get quarter map from month', () => {
        expect(() => getQuarterFromMonth(-1)).toThrow(TypeError);
        expect(() => getQuarterFromMonth(0)).toThrow(TypeError);
        expect(() => getQuarterFromMonth(14)).toThrow(TypeError);

        let quarter = getQuarterFromMonth(2);
        expect(quarter).toBe(1);

        quarter = getQuarterFromMonth(5);
        expect(quarter).toBe(2);

        quarter = getQuarterFromMonth(7);
        expect(quarter).toBe(3);

        quarter = getQuarterFromMonth(12);
        expect(quarter).toBe(4);
    })

    test('should calculate and return a list of total sales (sum) for each quarter', () => {
        quarterSumItems = sumSalesByQuarter(mySaleItems);
        expect(quarterSumItems).toMatchSnapshot();
    })

    test('should initial a quarterSumItem', () => {
        let newQuarterSumItems: QuarterSumItem[] = [];
        let quarterSumItemNew: QuarterSumItem = initQuarterSumItem(1, newQuarterSumItems);
        expect(quarterSumItemNew).toEqual({
            quarter: 1,
            totalPrices: 0,
            transactionNums: 0
        })
    })

    test('should get a quarterSumItem', () => {
        let quarterSumItemOld = getQuarterSumItem(1, quarterSumItems);
        expect(quarterSumItemOld).toMatchSnapshot();
    })

    test('should calculate and return a list of average sales for each quarter', () => {
        let quarterAverageItems: QuarterAverageItem[] = getQuarterAverageFrom(quarterSumItems);
        expect(quarterAverageItems).toMatchSnapshot();
    })
})