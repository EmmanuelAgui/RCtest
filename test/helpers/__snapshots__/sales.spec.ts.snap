// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers:sales get sales sorted by month should get sales sorted by a specified key ascend 1`] = `
Array [
  Object {
    "date": 3,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 4,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 1,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 5,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 5,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 7,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 9,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
  Object {
    "date": 3,
    "month": 11,
    "salePrice": 12,
    "transationId": "sdfsdf",
  },
]
`;

exports[`helpers:sales should calculate and return a list of average sales for each quarter 1`] = `
Array [
  Object {
    "averagePrices": 12,
    "quarter": 1,
    "transactionNums": 7,
  },
  Object {
    "averagePrices": 12,
    "quarter": 2,
    "transactionNums": 2,
  },
  Object {
    "averagePrices": 12,
    "quarter": 3,
    "transactionNums": 2,
  },
  Object {
    "averagePrices": 12,
    "quarter": 4,
    "transactionNums": 1,
  },
]
`;

exports[`helpers:sales should calculate and return a list of total sales (sum) for each quarter 1`] = `
Array [
  Object {
    "quarter": 1,
    "totalPrices": 84,
    "transactionNums": 7,
  },
  Object {
    "quarter": 2,
    "totalPrices": 24,
    "transactionNums": 2,
  },
  Object {
    "quarter": 3,
    "totalPrices": 24,
    "transactionNums": 2,
  },
  Object {
    "quarter": 4,
    "totalPrices": 12,
    "transactionNums": 1,
  },
]
`;

exports[`helpers:sales should get a quarterSumItem 1`] = `
Object {
  "quarter": 1,
  "totalPrices": 84,
  "transactionNums": 7,
}
`;
